eppo_core/src/event_ingestion/event_dispatcher.rs

```diff
@@ -0,0 +1,222 @@
+use crate::event_ingestion::batched_message::BatchedMessage;
+use crate::event_ingestion::event::Event;
+use crate::event_ingestion::event_delivery::EventDelivery;
+use crate::event_ingestion::queued_event::QueuedEvent;
+use crate::event_ingestion::retry::FinishedBatch;
+use crate::event_ingestion::{auto_flusher, batcher, delivery, retry};
+use tokio::sync::mpsc;
+use tokio::sync::mpsc::{Receiver, Sender};
+use tokio::time::Duration;
+use url::Url;
+
+// batch size of one means each event will be delivered individually, thus effectively disabling batching.
+const MIN_BATCH_SIZE: usize = 1;
+const MAX_BATCH_SIZE: usize = 10_000;
+
+#[derive(Debug, Clone)]
+pub(super) struct EventDispatcherConfig {
+    pub sdk_key: String,
+    pub ingestion_url: String,
+    pub delivery_interval: Duration,
+    pub retry_interval: Duration,
+    pub max_retry_delay: Duration,
+    pub max_retries: u8,
+    pub batch_size: usize,
+    pub max_queue_size: usize,
+}
+
+/// EventDispatcher is responsible for batching events and delivering them to the ingestion service
+/// via [`EventDelivery`].
+pub(super) struct EventDispatcher {
+    config: EventDispatcherConfig,
+}
+
+impl EventDispatcher {
+    pub fn new(config: EventDispatcherConfig) -> Self {
+        EventDispatcher { config }
+    }
+
+    /// Starts the event dispatcher related tasks and returns a sender and receiver pair.
+    /// Use the sender to dispatch events and the receiver to receive delivery statuses.
+    fn spawn_event_dispatcher(
+        &self,
+    ) -> (Sender<BatchedMessage<QueuedEvent>>, Receiver<FinishedBatch>) {
+        let config = self.config.clone();
+        let ingestion_url = Url::parse(config.ingestion_url.as_str())
+            .expect("Failed to create EventDelivery. invalid ingestion URL");
+        let event_delivery = EventDelivery::new(config.sdk_key.into(), ingestion_url);
+
+        // TODO: Does this make sense for channel size?
```

Comment:
Somewhat. I think it's OK for the first inbound channel, so we can buffer up to a batch of events from the user in case background thread is busy. (max_queue_size` seems more fitting though.)

For other channels, that would add extra buffers between all components which we generally don't want. Especially after batcher which compresses multiple events into one message, so `batch_size` channel after batcher would allow buffering up to `batch_size^2` events (`batch_size` messages each carrying `batch_size` events).

I would say that the rest of the channels should be size 1. (0/rendezvous would be preferred but tokio requires at least 1 buffer.)